design-docs/otel-collector-processors/README.md

```diff
@@ -0,0 +1,250 @@
+
+Virtual fields can be defined for the `type` of a signal (`span`, `metric`, `log`, `resource`) and the resource for a
+telemetry signal. For metrics, the structure presented for processing is actual data points, e.g. `NumberDataPoint`, 
+`HistogramDataPoint`, with the information from higher levels like `Metric` or the data type available as virtual fields.
+
+Navigation can then be used with a simple expression language for identifying telemetry to operate on.
+
+`... where name = "GET /cats"`
+`... where type = span and attributes["http.target"] = "/health"`
+`... where resource.attributes["deployment"] = "canary"`
+`... where type = metric and descriptor.metric_type = gauge`
+`... where type = metric and descriptor.metric_name = "http.active_requests"`
+
+Having selected telemetry to operate on, any needed operations can be defined as functions. Known useful functions should
+be implemented within the collector itself, provide registration from extension modules to allow customization with
+contrib components, and in the future can even allow user plugins possibly through WASM, similar to work in 
+[HTTP proxies](https://github.com/proxy-wasm/spec). The arguments to operations will primarily be field expressions,
+allowing the operation to mutate telemetry as needed.
+
+### Examples
+
+Remove a forbidden attribute such as `http.request.header.authorization` from all telemetry.
+
+`delete(attributes["http.request.header.authorization"])`
+
+Remove a forbidden attribute from spans only
+
+`delete(attributes["http.request.header.authorization"]) where type = span`
+
+Remove all attributes except for some
+
+`keep(attributes, "http.method", "http.status_code") where type = metric`
+
+Reduce cardinality of an attribute
+
+`replace_wildcards("/user/*/list/*", "/user/{userId}/list/{listId}", attributes["http.target"])`
+
+Reduce cardinality of a span name
+
+`replace_wildcards("GET /user/*/list/*", "GET /user/{userId}/list/{listId}", name) where type = span`
+
+Decrease the size of the telemetry payload by removing large resource attributes
+
+`delete(attributes["process.command_line"]) where type = resource)`
+
+Filtering out signals such as by removing all telemetry with a `http.target` of `/health`
+
+`drop() where attributes["http.target"] = "/health"`
+
+Attach information from resource into telemetry, for example adding certain resource fields as metric attributes
+
+`set(attributes["k8s_pod"], resource.attributes["k8s.pod.name"]) where type = metric`
+
+Stateful processing can also be modeled by the language. The processor implementation would set up the state while
+parsing the configuration.
+
+Create duration_metric with two attributes copied from a span
+
+```
+create_histogram("duration", end_time_nanos - start_time_nanos) where type = span
+keep(attributes, "http.method") where type = metric and descriptor.metric_name = "duration
+```
+
+Group spans by trace ID
```

Comment:
What does this do? I.e., are the grouped traces aggregated in some way or dumped out as `Pair<TraceId, List<Span>>`? What does the `2m` do?